In [1]:
import torch
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision import models
import torch.nn as nn
import torch.optim as optim
from triplet_utils import *
from utils import *
from torch.utils.data import random_split
import optuna
import torch

/home/sadat/miniconda3/envs/cowbytes/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# TripletModel definition
class TripletModel(nn.Module):
    def __init__(self):
        super(TripletModel, self).__init__()
        self.base_model = models.resnet50(pretrained=True)
        self.base_model.fc = nn.Linear(self.base_model.fc.in_features, 128)  # Adjust output size for embedding

    def forward(self, x):
        # Get the embedding directly from the base model
        return self.base_model(x)  # Call the model with input x

class TripletClassifier(nn.Module):
    def __init__(self, triplet_head, output_size=1):
        super(TripletClassifier, self).__init__()
        self.triplet_head = triplet_head
        self.classifier = nn.Linear(self.triplet_head.base_model.fc.out_features, output_size)  # Binary classification (logits)

    def forward(self, x):
        embedding = self.triplet_head(x)  # Output of base model
        class_output = self.classifier(embedding)  # Output for classification
        return embedding, class_output

In [3]:
def custom_collate(batch):
    # Unpack the batch
    anchors, positives, negatives, labels = zip(*batch)

    # Stack tensors to create batches
    anchors = torch.stack(anchors)
    positives = torch.stack(positives)
    negatives = torch.stack(negatives)

    # Convert labels to a tensor
    labels = torch.tensor(labels, dtype=torch.float32).unsqueeze(1)

    return anchors, positives, negatives, labels

In [4]:
# Load data
data_dir = '../../data/BiteCount/salient_poses/'

# Define transformations for training set and validation set
train_transforms = transforms.Compose([
    transforms.Resize((224, 224)),  # Resizing to square input
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
])

val_transforms = transforms.Compose([
    transforms.Resize((224, 224)),  # Resizing to square input
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
])

# Load dataset
full_dataset = TripletDataset(data_dir, transform=train_transforms)

# Define the split ratio
train_size = int(0.8 * len(full_dataset))  # 80% for training
val_size = len(full_dataset) - train_size  # 20% for validation

# Split the dataset into training and validation sets
train_dataset, val_dataset = random_split(full_dataset, [train_size, val_size])

# Create DataLoader for train and validation
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True, collate_fn=custom_collate)
val_loader = DataLoader(val_dataset, batch_size=16, shuffle=False, collate_fn=custom_collate)

def objective(trial):
    # Suggest hyperparameters
    learning_rate = trial.suggest_loguniform('learning_rate', 1e-5, 1e-1)
    momentum = trial.suggest_uniform('momentum', 0.5, 0.99)
    triplet_margin = trial.suggest_uniform('triplet_margin', 0.1, 2.0)
    classification_loss_weight = trial.suggest_uniform('classification_loss_weight', 0.1, 1.0)

    print(f"\nTrial {trial.number}:")
    print(f"Suggested Hyperparameters: learning_rate={learning_rate}, momentum={momentum}, "
          f"triplet_margin={triplet_margin}, classification_loss_weight={classification_loss_weight}")
    
    # Initialize model
    triplet_model = TripletModel()
    model = TripletClassifier(triplet_model)
    model = model.to(device)

    # Define loss functions and optimizer
    criterion_classification = nn.BCEWithLogitsLoss()
    criterion_triplet = nn.TripletMarginLoss(margin=triplet_margin)
    optimizer = optim.SGD(model.parameters(), lr=learning_rate, momentum=momentum)

    # Training loop
    num_epochs = 25
    best_epoch_val_loss = float('inf')

    for epoch in range(num_epochs):
        model.train()
        running_loss = 0.0

        for batch in train_loader:
            # Unpack the batch correctly
            anchor, positive, negative, labels = batch
            
            # Move the tensors to the appropriate device
            anchor = anchor.to(device)
            positive = positive.to(device)
            negative = negative.to(device)
            labels = labels.to(device)  # Ensure labels are tensors

            # Forward pass
            anchor_embedding, anchor_output = model(anchor)
            positive_embedding, _ = model(positive)
            negative_embedding, _ = model(negative)

            # Calculate losses
            classification_loss = criterion_classification(anchor_output, labels)
            triplet_loss = criterion_triplet(anchor_embedding, positive_embedding, negative_embedding)

            # Combine losses with weighting
            loss = classification_loss_weight * classification_loss + triplet_loss
            running_loss += loss.item()

            # Backward pass and optimization
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        # Validation step
        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for batch in val_loader:
                anchor, positive, negative, labels = batch
                anchor = anchor.to(device)
                positive = positive.to(device)
                negative = negative.to(device)
                labels = labels.to(device)

                anchor_embedding, anchor_output = model(anchor)
                positive_embedding, _ = model(positive)
                negative_embedding, _ = model(negative)
                classification_loss = criterion_classification(anchor_output, labels)
                triplet_loss = criterion_triplet(anchor_embedding, positive_embedding, negative_embedding)

                val_loss += (classification_loss).item()

        avg_val_loss = val_loss / len(val_loader)

        # Check for improvement
        if avg_val_loss < best_epoch_val_loss:
            best_epoch_val_loss = avg_val_loss

        print(f"Epoch {epoch + 1}/{num_epochs}: "
              f"Validation Loss: {avg_val_loss:.4f}")

    return best_epoch_val_loss  # Return the average validation loss

# Create a study object and optimize the objective function
study = optuna.create_study(direction='minimize')
study.optimize(objective, n_trials=30)  # Adjust the number of trials as needed

# Print the best hyperparameters and validation loss
print("Best hyperparameters: ", study.best_params)
print("Best validation loss: ", study.best_value)

[I 2024-10-03 19:39:18,676] A new study created in memory with name: no-name-6e1db1cf-099c-484a-aaf1-1d56421fe055
/tmp/ipykernel_25998/3782639035.py:34: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  learning_rate = trial.suggest_loguniform('learning_rate', 1e-5, 1e-1)
/tmp/ipykernel_25998/3782639035.py:35: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  momentum = trial.suggest_uniform('momentum', 0.5, 0.99)
/tmp/ipykernel_25998/3782639035.py:36: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  triplet_margin = trial.suggest_uniform('triplet_margin', 0


Trial 0:
Suggested Hyperparameters: learning_rate=0.00010513056948022352, momentum=0.7687324015832673, triplet_margin=0.5734419515063479, classification_loss_weight=0.42324119802267746
Epoch 1/25: Validation Loss: 0.7032
Epoch 2/25: Validation Loss: 0.7043
Epoch 3/25: Validation Loss: 0.7068
Epoch 4/25: Validation Loss: 0.6995
Epoch 5/25: Validation Loss: 0.7013
Epoch 6/25: Validation Loss: 0.7032
Epoch 7/25: Validation Loss: 0.7052
Epoch 8/25: Validation Loss: 0.7063
Epoch 9/25: Validation Loss: 0.7010
Epoch 10/25: Validation Loss: 0.7003
Epoch 11/25: Validation Loss: 0.7007
Epoch 12/25: Validation Loss: 0.7062
Epoch 13/25: Validation Loss: 0.7006
Epoch 14/25: Validation Loss: 0.6985
Epoch 15/25: Validation Loss: 0.7026
Epoch 16/25: Validation Loss: 0.6973
Epoch 17/25: Validation Loss: 0.6951
Epoch 18/25: Validation Loss: 0.7027
Epoch 19/25: Validation Loss: 0.6990
Epoch 20/25: Validation Loss: 0.6964
Epoch 21/25: Validation Loss: 0.7030
Epoch 22/25: Validation Loss: 0.6949
Epoch 23/

[I 2024-10-03 19:41:27,406] Trial 0 finished with value: 0.6948638677597045 and parameters: {'learning_rate': 0.00010513056948022352, 'momentum': 0.7687324015832673, 'triplet_margin': 0.5734419515063479, 'classification_loss_weight': 0.42324119802267746}. Best is trial 0 with value: 0.6948638677597045.


Epoch 25/25: Validation Loss: 0.6971

Trial 1:
Suggested Hyperparameters: learning_rate=0.0005600848988312334, momentum=0.6327792322084209, triplet_margin=0.6189066125747088, classification_loss_weight=0.5434323177802199
Epoch 1/25: Validation Loss: 0.6977
Epoch 2/25: Validation Loss: 0.6972
Epoch 3/25: Validation Loss: 0.6996
Epoch 4/25: Validation Loss: 0.7032
Epoch 5/25: Validation Loss: 0.6966
Epoch 6/25: Validation Loss: 0.6984
Epoch 7/25: Validation Loss: 0.6915
Epoch 8/25: Validation Loss: 0.6982
Epoch 9/25: Validation Loss: 0.6996
Epoch 10/25: Validation Loss: 0.6983
Epoch 11/25: Validation Loss: 0.6935
Epoch 12/25: Validation Loss: 0.6974
Epoch 13/25: Validation Loss: 0.6998
Epoch 14/25: Validation Loss: 0.6965
Epoch 15/25: Validation Loss: 0.6930
Epoch 16/25: Validation Loss: 0.6987
Epoch 17/25: Validation Loss: 0.6949
Epoch 18/25: Validation Loss: 0.6968
Epoch 19/25: Validation Loss: 0.6973
Epoch 20/25: Validation Loss: 0.6970
Epoch 21/25: Validation Loss: 0.6963
Epoch 22/25

[I 2024-10-03 19:43:28,552] Trial 1 finished with value: 0.6914852499961853 and parameters: {'learning_rate': 0.0005600848988312334, 'momentum': 0.6327792322084209, 'triplet_margin': 0.6189066125747088, 'classification_loss_weight': 0.5434323177802199}. Best is trial 1 with value: 0.6914852499961853.


Epoch 25/25: Validation Loss: 0.6953

Trial 2:
Suggested Hyperparameters: learning_rate=1.8113751286127613e-05, momentum=0.7008303717757595, triplet_margin=1.8409021136058987, classification_loss_weight=0.5494288313587101
Epoch 1/25: Validation Loss: 0.6971
Epoch 2/25: Validation Loss: 0.6910
Epoch 3/25: Validation Loss: 0.6916
Epoch 4/25: Validation Loss: 0.6944
Epoch 5/25: Validation Loss: 0.6997
Epoch 6/25: Validation Loss: 0.6933
Epoch 7/25: Validation Loss: 0.7010
Epoch 8/25: Validation Loss: 0.6967
Epoch 9/25: Validation Loss: 0.6963
Epoch 10/25: Validation Loss: 0.6917
Epoch 11/25: Validation Loss: 0.6968
Epoch 12/25: Validation Loss: 0.6950
Epoch 13/25: Validation Loss: 0.6986
Epoch 14/25: Validation Loss: 0.6876
Epoch 15/25: Validation Loss: 0.6941
Epoch 16/25: Validation Loss: 0.6964
Epoch 17/25: Validation Loss: 0.6903
Epoch 18/25: Validation Loss: 0.6944
Epoch 19/25: Validation Loss: 0.6938
Epoch 20/25: Validation Loss: 0.6929
Epoch 21/25: Validation Loss: 0.6908
Epoch 22/2

[I 2024-10-03 19:45:32,645] Trial 2 finished with value: 0.6875911712646484 and parameters: {'learning_rate': 1.8113751286127613e-05, 'momentum': 0.7008303717757595, 'triplet_margin': 1.8409021136058987, 'classification_loss_weight': 0.5494288313587101}. Best is trial 2 with value: 0.6875911712646484.


Epoch 25/25: Validation Loss: 0.6975

Trial 3:
Suggested Hyperparameters: learning_rate=0.05384918169864502, momentum=0.9199121038740428, triplet_margin=0.4239308409097565, classification_loss_weight=0.7752549169953352
Epoch 1/25: Validation Loss: 0.7228
Epoch 2/25: Validation Loss: 0.9174
Epoch 3/25: Validation Loss: 0.7276
Epoch 4/25: Validation Loss: 0.7000
Epoch 5/25: Validation Loss: 0.6995
Epoch 6/25: Validation Loss: 0.6919
Epoch 7/25: Validation Loss: 0.6943
Epoch 8/25: Validation Loss: 0.6981
Epoch 9/25: Validation Loss: 0.6950
Epoch 10/25: Validation Loss: 0.6919
Epoch 11/25: Validation Loss: 0.6946
Epoch 12/25: Validation Loss: 0.6925
Epoch 13/25: Validation Loss: 0.7038
Epoch 14/25: Validation Loss: 0.6963
Epoch 15/25: Validation Loss: 0.6919
Epoch 16/25: Validation Loss: 0.6925
Epoch 17/25: Validation Loss: 0.6941
Epoch 18/25: Validation Loss: 0.6938
Epoch 19/25: Validation Loss: 0.6924
Epoch 20/25: Validation Loss: 0.6945
Epoch 21/25: Validation Loss: 0.6989
Epoch 22/25: 

[I 2024-10-03 19:47:36,425] Trial 3 finished with value: 0.6918790221214295 and parameters: {'learning_rate': 0.05384918169864502, 'momentum': 0.9199121038740428, 'triplet_margin': 0.4239308409097565, 'classification_loss_weight': 0.7752549169953352}. Best is trial 2 with value: 0.6875911712646484.


Epoch 25/25: Validation Loss: 0.6962

Trial 4:
Suggested Hyperparameters: learning_rate=0.00012417025855222577, momentum=0.7720503209281782, triplet_margin=0.5639955636907779, classification_loss_weight=0.405292637630718
Epoch 1/25: Validation Loss: 0.6925
Epoch 2/25: Validation Loss: 0.6956
Epoch 3/25: Validation Loss: 0.6886
Epoch 4/25: Validation Loss: 0.6927
Epoch 5/25: Validation Loss: 0.6970
Epoch 6/25: Validation Loss: 0.6969
Epoch 7/25: Validation Loss: 0.6990
Epoch 8/25: Validation Loss: 0.6962
Epoch 9/25: Validation Loss: 0.6974
Epoch 10/25: Validation Loss: 0.6931
Epoch 11/25: Validation Loss: 0.6995
Epoch 12/25: Validation Loss: 0.6869
Epoch 13/25: Validation Loss: 0.6960
Epoch 14/25: Validation Loss: 0.6940
Epoch 15/25: Validation Loss: 0.6923
Epoch 16/25: Validation Loss: 0.6941
Epoch 17/25: Validation Loss: 0.6960
Epoch 18/25: Validation Loss: 0.6930
Epoch 19/25: Validation Loss: 0.6925
Epoch 20/25: Validation Loss: 0.6897
Epoch 21/25: Validation Loss: 0.6928
Epoch 22/25

[I 2024-10-03 19:49:38,222] Trial 4 finished with value: 0.6869475722312928 and parameters: {'learning_rate': 0.00012417025855222577, 'momentum': 0.7720503209281782, 'triplet_margin': 0.5639955636907779, 'classification_loss_weight': 0.405292637630718}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6918

Trial 5:
Suggested Hyperparameters: learning_rate=0.0009334030050773037, momentum=0.7786798310284739, triplet_margin=0.15446600783435607, classification_loss_weight=0.5986148641338181
Epoch 1/25: Validation Loss: 0.7039
Epoch 2/25: Validation Loss: 0.7011
Epoch 3/25: Validation Loss: 0.6987
Epoch 4/25: Validation Loss: 0.6978
Epoch 5/25: Validation Loss: 0.6971
Epoch 6/25: Validation Loss: 0.6983
Epoch 7/25: Validation Loss: 0.6960
Epoch 8/25: Validation Loss: 0.6993
Epoch 9/25: Validation Loss: 0.7002
Epoch 10/25: Validation Loss: 0.6972
Epoch 11/25: Validation Loss: 0.6970
Epoch 12/25: Validation Loss: 0.6984
Epoch 13/25: Validation Loss: 0.6990
Epoch 14/25: Validation Loss: 0.6929
Epoch 15/25: Validation Loss: 0.6964
Epoch 16/25: Validation Loss: 0.6961
Epoch 17/25: Validation Loss: 0.6938
Epoch 18/25: Validation Loss: 0.6996
Epoch 19/25: Validation Loss: 0.6989
Epoch 20/25: Validation Loss: 0.6974
Epoch 21/25: Validation Loss: 0.6954
Epoch 22/2

[I 2024-10-03 19:51:41,200] Trial 5 finished with value: 0.6884064316749573 and parameters: {'learning_rate': 0.0009334030050773037, 'momentum': 0.7786798310284739, 'triplet_margin': 0.15446600783435607, 'classification_loss_weight': 0.5986148641338181}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6951

Trial 6:
Suggested Hyperparameters: learning_rate=0.00016075560306593928, momentum=0.5807546192658697, triplet_margin=1.7727458319858111, classification_loss_weight=0.18098933214482943
Epoch 1/25: Validation Loss: 0.6930
Epoch 2/25: Validation Loss: 0.6978
Epoch 3/25: Validation Loss: 0.6909
Epoch 4/25: Validation Loss: 0.6997
Epoch 5/25: Validation Loss: 0.6970
Epoch 6/25: Validation Loss: 0.6935
Epoch 7/25: Validation Loss: 0.6942
Epoch 8/25: Validation Loss: 0.6976
Epoch 9/25: Validation Loss: 0.6918
Epoch 10/25: Validation Loss: 0.6916
Epoch 11/25: Validation Loss: 0.6998
Epoch 12/25: Validation Loss: 0.6959
Epoch 13/25: Validation Loss: 0.6935
Epoch 14/25: Validation Loss: 0.6928
Epoch 15/25: Validation Loss: 0.6974
Epoch 16/25: Validation Loss: 0.6981
Epoch 17/25: Validation Loss: 0.6976
Epoch 18/25: Validation Loss: 0.6929
Epoch 19/25: Validation Loss: 0.6931
Epoch 20/25: Validation Loss: 0.6925
Epoch 21/25: Validation Loss: 0.6912
Epoch 22/

[I 2024-10-03 19:53:45,085] Trial 6 finished with value: 0.690859067440033 and parameters: {'learning_rate': 0.00016075560306593928, 'momentum': 0.5807546192658697, 'triplet_margin': 1.7727458319858111, 'classification_loss_weight': 0.18098933214482943}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6964

Trial 7:
Suggested Hyperparameters: learning_rate=6.183691602869704e-05, momentum=0.902015607761429, triplet_margin=1.9696016150631985, classification_loss_weight=0.6896500050089751
Epoch 1/25: Validation Loss: 0.6897
Epoch 2/25: Validation Loss: 0.6877
Epoch 3/25: Validation Loss: 0.6940
Epoch 4/25: Validation Loss: 0.6933
Epoch 5/25: Validation Loss: 0.6946
Epoch 6/25: Validation Loss: 0.6947
Epoch 7/25: Validation Loss: 0.6943
Epoch 8/25: Validation Loss: 0.6904
Epoch 9/25: Validation Loss: 0.6939
Epoch 10/25: Validation Loss: 0.6917
Epoch 11/25: Validation Loss: 0.6965
Epoch 12/25: Validation Loss: 0.6940
Epoch 13/25: Validation Loss: 0.6885
Epoch 14/25: Validation Loss: 0.6962
Epoch 15/25: Validation Loss: 0.6934
Epoch 16/25: Validation Loss: 0.6901
Epoch 17/25: Validation Loss: 0.6908
Epoch 18/25: Validation Loss: 0.6960
Epoch 19/25: Validation Loss: 0.6939
Epoch 20/25: Validation Loss: 0.6959
Epoch 21/25: Validation Loss: 0.6927
Epoch 22/25:

[I 2024-10-03 19:55:49,155] Trial 7 finished with value: 0.687668776512146 and parameters: {'learning_rate': 6.183691602869704e-05, 'momentum': 0.902015607761429, 'triplet_margin': 1.9696016150631985, 'classification_loss_weight': 0.6896500050089751}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6913

Trial 8:
Suggested Hyperparameters: learning_rate=0.05441649495055668, momentum=0.975447227975866, triplet_margin=0.9524345480465405, classification_loss_weight=0.7510973489998745
Epoch 1/25: Validation Loss: 0.8035
Epoch 2/25: Validation Loss: 1.9341
Epoch 3/25: Validation Loss: 1.3948
Epoch 4/25: Validation Loss: 0.7000
Epoch 5/25: Validation Loss: 0.8950
Epoch 6/25: Validation Loss: 0.6927
Epoch 7/25: Validation Loss: 1.0940
Epoch 8/25: Validation Loss: 0.7009
Epoch 9/25: Validation Loss: 0.7106
Epoch 10/25: Validation Loss: 0.6917
Epoch 11/25: Validation Loss: 0.7085
Epoch 12/25: Validation Loss: 0.7452
Epoch 13/25: Validation Loss: 0.6923
Epoch 14/25: Validation Loss: 0.6936
Epoch 15/25: Validation Loss: 0.7062
Epoch 16/25: Validation Loss: 0.6922
Epoch 17/25: Validation Loss: 0.6965
Epoch 18/25: Validation Loss: 0.7005
Epoch 19/25: Validation Loss: 0.6929
Epoch 20/25: Validation Loss: 0.6920
Epoch 21/25: Validation Loss: 0.6950
Epoch 22/25: V

[I 2024-10-03 19:57:51,423] Trial 8 finished with value: 0.6917022109031677 and parameters: {'learning_rate': 0.05441649495055668, 'momentum': 0.975447227975866, 'triplet_margin': 0.9524345480465405, 'classification_loss_weight': 0.7510973489998745}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6919

Trial 9:
Suggested Hyperparameters: learning_rate=0.09783018043813542, momentum=0.606888006227554, triplet_margin=1.9153142781722443, classification_loss_weight=0.8066067566777212
Epoch 1/25: Validation Loss: 0.8949
Epoch 2/25: Validation Loss: 0.6990
Epoch 3/25: Validation Loss: 0.6943
Epoch 4/25: Validation Loss: 0.6939
Epoch 5/25: Validation Loss: 0.6911
Epoch 6/25: Validation Loss: 0.6944
Epoch 7/25: Validation Loss: 0.6961
Epoch 8/25: Validation Loss: 0.6944
Epoch 9/25: Validation Loss: 0.6934
Epoch 10/25: Validation Loss: 0.6933
Epoch 11/25: Validation Loss: 0.6934
Epoch 12/25: Validation Loss: 0.7009
Epoch 13/25: Validation Loss: 0.6958
Epoch 14/25: Validation Loss: 0.6958
Epoch 15/25: Validation Loss: 0.6960
Epoch 16/25: Validation Loss: 0.6934
Epoch 17/25: Validation Loss: 0.6923
Epoch 18/25: Validation Loss: 0.6932
Epoch 19/25: Validation Loss: 0.6937
Epoch 20/25: Validation Loss: 0.6941
Epoch 21/25: Validation Loss: 0.6927
Epoch 22/25: V

[I 2024-10-03 19:59:55,040] Trial 9 finished with value: 0.6910863757133484 and parameters: {'learning_rate': 0.09783018043813542, 'momentum': 0.606888006227554, 'triplet_margin': 1.9153142781722443, 'classification_loss_weight': 0.8066067566777212}. Best is trial 4 with value: 0.6869475722312928.


Epoch 25/25: Validation Loss: 0.6941

Trial 10:
Suggested Hyperparameters: learning_rate=0.00730034646214492, momentum=0.8086092111230203, triplet_margin=1.3118824464954721, classification_loss_weight=0.9835969240329065
Epoch 1/25: Validation Loss: 0.6983
Epoch 2/25: Validation Loss: 0.6925
Epoch 3/25: Validation Loss: 0.6928
Epoch 4/25: Validation Loss: 0.6965
Epoch 5/25: Validation Loss: 0.6764
Epoch 6/25: Validation Loss: 0.6862
Epoch 7/25: Validation Loss: 0.6685
Epoch 8/25: Validation Loss: 0.6349
Epoch 9/25: Validation Loss: 0.5645
Epoch 10/25: Validation Loss: 0.5827
Epoch 11/25: Validation Loss: 0.5671
Epoch 12/25: Validation Loss: 0.5843
Epoch 13/25: Validation Loss: 0.4450
Epoch 14/25: Validation Loss: 0.4593
Epoch 15/25: Validation Loss: 0.4062
Epoch 16/25: Validation Loss: 0.3933
Epoch 17/25: Validation Loss: 0.4124
Epoch 18/25: Validation Loss: 0.3505
Epoch 19/25: Validation Loss: 0.4632
Epoch 20/25: Validation Loss: 0.4036
Epoch 21/25: Validation Loss: 0.3966
Epoch 22/25:

[I 2024-10-03 20:01:56,150] Trial 10 finished with value: 0.2511402040719986 and parameters: {'learning_rate': 0.00730034646214492, 'momentum': 0.8086092111230203, 'triplet_margin': 1.3118824464954721, 'classification_loss_weight': 0.9835969240329065}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.2825

Trial 11:
Suggested Hyperparameters: learning_rate=0.006007707570969162, momentum=0.8278002508597458, triplet_margin=1.3474915653389872, classification_loss_weight=0.3017360365522086
Epoch 1/25: Validation Loss: 0.7021
Epoch 2/25: Validation Loss: 0.6953
Epoch 3/25: Validation Loss: 0.6959
Epoch 4/25: Validation Loss: 0.6904
Epoch 5/25: Validation Loss: 0.6901
Epoch 6/25: Validation Loss: 0.6980
Epoch 7/25: Validation Loss: 0.6881
Epoch 8/25: Validation Loss: 0.6892
Epoch 9/25: Validation Loss: 0.6787
Epoch 10/25: Validation Loss: 0.6714
Epoch 11/25: Validation Loss: 0.6575
Epoch 12/25: Validation Loss: 0.6197
Epoch 13/25: Validation Loss: 0.6216
Epoch 14/25: Validation Loss: 0.6037
Epoch 15/25: Validation Loss: 0.5473
Epoch 16/25: Validation Loss: 0.5361
Epoch 17/25: Validation Loss: 0.5187
Epoch 18/25: Validation Loss: 0.4548
Epoch 19/25: Validation Loss: 0.4512
Epoch 20/25: Validation Loss: 0.4222
Epoch 21/25: Validation Loss: 0.3726
Epoch 22/25

[I 2024-10-03 20:03:59,144] Trial 11 finished with value: 0.34909220337867736 and parameters: {'learning_rate': 0.006007707570969162, 'momentum': 0.8278002508597458, 'triplet_margin': 1.3474915653389872, 'classification_loss_weight': 0.3017360365522086}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.3491

Trial 12:
Suggested Hyperparameters: learning_rate=0.0068769412749483585, momentum=0.8399870180508793, triplet_margin=1.4906971122139456, classification_loss_weight=0.9415002388025169
Epoch 1/25: Validation Loss: 0.7001
Epoch 2/25: Validation Loss: 0.6902
Epoch 3/25: Validation Loss: 0.6996
Epoch 4/25: Validation Loss: 0.6944
Epoch 5/25: Validation Loss: 0.6968
Epoch 6/25: Validation Loss: 0.6765
Epoch 7/25: Validation Loss: 0.6561
Epoch 8/25: Validation Loss: 0.6283
Epoch 9/25: Validation Loss: 0.6618
Epoch 10/25: Validation Loss: 0.6157
Epoch 11/25: Validation Loss: 0.4601
Epoch 12/25: Validation Loss: 0.4911
Epoch 13/25: Validation Loss: 0.5428
Epoch 14/25: Validation Loss: 0.4811
Epoch 15/25: Validation Loss: 0.4760
Epoch 16/25: Validation Loss: 0.3771
Epoch 17/25: Validation Loss: 0.4904
Epoch 18/25: Validation Loss: 0.3614
Epoch 19/25: Validation Loss: 0.2977
Epoch 20/25: Validation Loss: 0.3583
Epoch 21/25: Validation Loss: 0.4420
Epoch 22/2

[I 2024-10-03 20:06:01,394] Trial 12 finished with value: 0.2977206587791443 and parameters: {'learning_rate': 0.0068769412749483585, 'momentum': 0.8399870180508793, 'triplet_margin': 1.4906971122139456, 'classification_loss_weight': 0.9415002388025169}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.3306

Trial 13:
Suggested Hyperparameters: learning_rate=0.00802703280930498, momentum=0.8558159428190063, triplet_margin=1.4127408450467325, classification_loss_weight=0.9835123659777987
Epoch 1/25: Validation Loss: 0.6945
Epoch 2/25: Validation Loss: 0.7023
Epoch 3/25: Validation Loss: 0.6976
Epoch 4/25: Validation Loss: 0.6935
Epoch 5/25: Validation Loss: 0.6883
Epoch 6/25: Validation Loss: 0.6796
Epoch 7/25: Validation Loss: 0.6173
Epoch 8/25: Validation Loss: 0.5589
Epoch 9/25: Validation Loss: 0.6207
Epoch 10/25: Validation Loss: 0.5734
Epoch 11/25: Validation Loss: 0.4607
Epoch 12/25: Validation Loss: 0.5386
Epoch 13/25: Validation Loss: 0.4325
Epoch 14/25: Validation Loss: 0.4181
Epoch 15/25: Validation Loss: 0.4437
Epoch 16/25: Validation Loss: 0.4073
Epoch 17/25: Validation Loss: 0.4962
Epoch 18/25: Validation Loss: 0.4493
Epoch 19/25: Validation Loss: 0.3550
Epoch 20/25: Validation Loss: 0.4150
Epoch 21/25: Validation Loss: 0.2940
Epoch 22/25:

[I 2024-10-03 20:08:04,948] Trial 13 finished with value: 0.2940253585577011 and parameters: {'learning_rate': 0.00802703280930498, 'momentum': 0.8558159428190063, 'triplet_margin': 1.4127408450467325, 'classification_loss_weight': 0.9835123659777987}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.2985

Trial 14:
Suggested Hyperparameters: learning_rate=0.00592068669369857, momentum=0.6925625010506282, triplet_margin=1.145323648351063, classification_loss_weight=0.9692826078571686
Epoch 1/25: Validation Loss: 0.6901
Epoch 2/25: Validation Loss: 0.6964
Epoch 3/25: Validation Loss: 0.6973
Epoch 4/25: Validation Loss: 0.6935
Epoch 5/25: Validation Loss: 0.6884
Epoch 6/25: Validation Loss: 0.6957
Epoch 7/25: Validation Loss: 0.6879
Epoch 8/25: Validation Loss: 0.6897
Epoch 9/25: Validation Loss: 0.6867
Epoch 10/25: Validation Loss: 0.6681
Epoch 11/25: Validation Loss: 0.6766
Epoch 12/25: Validation Loss: 0.6624
Epoch 13/25: Validation Loss: 0.6330
Epoch 14/25: Validation Loss: 0.6588
Epoch 15/25: Validation Loss: 0.5665
Epoch 16/25: Validation Loss: 0.5805
Epoch 17/25: Validation Loss: 0.5105
Epoch 18/25: Validation Loss: 0.4505
Epoch 19/25: Validation Loss: 0.4320
Epoch 20/25: Validation Loss: 0.4065
Epoch 21/25: Validation Loss: 0.2967
Epoch 22/25: 

[I 2024-10-03 20:10:08,417] Trial 14 finished with value: 0.29434171617031096 and parameters: {'learning_rate': 0.00592068669369857, 'momentum': 0.6925625010506282, 'triplet_margin': 1.145323648351063, 'classification_loss_weight': 0.9692826078571686}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.2943

Trial 15:
Suggested Hyperparameters: learning_rate=0.01597058198656061, momentum=0.8515085666379474, triplet_margin=1.572438708342606, classification_loss_weight=0.8747388506940991
Epoch 1/25: Validation Loss: 0.6987
Epoch 2/25: Validation Loss: 0.7023
Epoch 3/25: Validation Loss: 0.7337
Epoch 4/25: Validation Loss: 0.6946
Epoch 5/25: Validation Loss: 0.7235
Epoch 6/25: Validation Loss: 0.6630
Epoch 7/25: Validation Loss: 0.6448
Epoch 8/25: Validation Loss: 0.6776
Epoch 9/25: Validation Loss: 0.6547
Epoch 10/25: Validation Loss: 0.6493
Epoch 11/25: Validation Loss: 0.5904
Epoch 12/25: Validation Loss: 0.6648
Epoch 13/25: Validation Loss: 0.5837
Epoch 14/25: Validation Loss: 0.5969
Epoch 15/25: Validation Loss: 0.6056
Epoch 16/25: Validation Loss: 0.5926
Epoch 17/25: Validation Loss: 0.6133
Epoch 18/25: Validation Loss: 0.8347
Epoch 19/25: Validation Loss: 0.6237
Epoch 20/25: Validation Loss: 0.5411
Epoch 21/25: Validation Loss: 0.4754
Epoch 22/25: 

[I 2024-10-03 20:12:10,515] Trial 15 finished with value: 0.33964827954769133 and parameters: {'learning_rate': 0.01597058198656061, 'momentum': 0.8515085666379474, 'triplet_margin': 1.572438708342606, 'classification_loss_weight': 0.8747388506940991}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.3396

Trial 16:
Suggested Hyperparameters: learning_rate=0.002918104486119395, momentum=0.9777373870964376, triplet_margin=1.0227338207056662, classification_loss_weight=0.8903787481139779
Epoch 1/25: Validation Loss: 0.6955
Epoch 2/25: Validation Loss: 0.6924
Epoch 3/25: Validation Loss: 0.7003
Epoch 4/25: Validation Loss: 0.6950
Epoch 5/25: Validation Loss: 0.6932
Epoch 6/25: Validation Loss: 0.6876
Epoch 7/25: Validation Loss: 0.6942
Epoch 8/25: Validation Loss: 0.6720
Epoch 9/25: Validation Loss: 0.6812
Epoch 10/25: Validation Loss: 0.6033
Epoch 11/25: Validation Loss: 0.6227
Epoch 12/25: Validation Loss: 0.6181
Epoch 13/25: Validation Loss: 0.5980
Epoch 14/25: Validation Loss: 0.5844
Epoch 15/25: Validation Loss: 0.5841
Epoch 16/25: Validation Loss: 0.5215
Epoch 17/25: Validation Loss: 0.5426
Epoch 18/25: Validation Loss: 0.4565
Epoch 19/25: Validation Loss: 0.5305
Epoch 20/25: Validation Loss: 0.4532
Epoch 21/25: Validation Loss: 0.5659
Epoch 22/25

[I 2024-10-03 20:14:11,357] Trial 16 finished with value: 0.4531998872756958 and parameters: {'learning_rate': 0.002918104486119395, 'momentum': 0.9777373870964376, 'triplet_margin': 1.0227338207056662, 'classification_loss_weight': 0.8903787481139779}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.4541

Trial 17:
Suggested Hyperparameters: learning_rate=0.01925286961725715, momentum=0.7035252987211111, triplet_margin=1.2856856981589964, classification_loss_weight=0.9758887386718365
Epoch 1/25: Validation Loss: 0.7162
Epoch 2/25: Validation Loss: 0.7032
Epoch 3/25: Validation Loss: 0.7137
Epoch 4/25: Validation Loss: 0.7003
Epoch 5/25: Validation Loss: 0.6980
Epoch 6/25: Validation Loss: 0.6613
Epoch 7/25: Validation Loss: 0.6784
Epoch 8/25: Validation Loss: 0.5894
Epoch 9/25: Validation Loss: 0.7123
Epoch 10/25: Validation Loss: 0.4315
Epoch 11/25: Validation Loss: 0.7984
Epoch 12/25: Validation Loss: 0.4866
Epoch 13/25: Validation Loss: 0.5316
Epoch 14/25: Validation Loss: 0.4464
Epoch 15/25: Validation Loss: 0.4962
Epoch 16/25: Validation Loss: 0.4264
Epoch 17/25: Validation Loss: 0.4933
Epoch 18/25: Validation Loss: 0.4361
Epoch 19/25: Validation Loss: 0.5373
Epoch 20/25: Validation Loss: 0.5000
Epoch 21/25: Validation Loss: 0.3988
Epoch 22/25:

[I 2024-10-03 20:16:14,703] Trial 17 finished with value: 0.33825955986976625 and parameters: {'learning_rate': 0.01925286961725715, 'momentum': 0.7035252987211111, 'triplet_margin': 1.2856856981589964, 'classification_loss_weight': 0.9758887386718365}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.3383

Trial 18:
Suggested Hyperparameters: learning_rate=0.0020304972902228154, momentum=0.5302113035070135, triplet_margin=0.8553954443759731, classification_loss_weight=0.6805970413156339
Epoch 1/25: Validation Loss: 0.7006
Epoch 2/25: Validation Loss: 0.6951
Epoch 3/25: Validation Loss: 0.7023
Epoch 4/25: Validation Loss: 0.6944
Epoch 5/25: Validation Loss: 0.6943
Epoch 6/25: Validation Loss: 0.6950
Epoch 7/25: Validation Loss: 0.6932
Epoch 8/25: Validation Loss: 0.6967
Epoch 9/25: Validation Loss: 0.6976
Epoch 10/25: Validation Loss: 0.6964
Epoch 11/25: Validation Loss: 0.6997
Epoch 12/25: Validation Loss: 0.6914
Epoch 13/25: Validation Loss: 0.6981
Epoch 14/25: Validation Loss: 0.6943
Epoch 15/25: Validation Loss: 0.6954
Epoch 16/25: Validation Loss: 0.6961
Epoch 17/25: Validation Loss: 0.6903
Epoch 18/25: Validation Loss: 0.7011
Epoch 19/25: Validation Loss: 0.6940
Epoch 20/25: Validation Loss: 0.6978
Epoch 21/25: Validation Loss: 0.6884
Epoch 22/2

[I 2024-10-03 20:18:15,962] Trial 18 finished with value: 0.6842079281806945 and parameters: {'learning_rate': 0.0020304972902228154, 'momentum': 0.5302113035070135, 'triplet_margin': 0.8553954443759731, 'classification_loss_weight': 0.6805970413156339}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.6950

Trial 19:
Suggested Hyperparameters: learning_rate=0.00044790758123736634, momentum=0.9008586002953491, triplet_margin=1.6185130873211857, classification_loss_weight=0.8356618591942357
Epoch 1/25: Validation Loss: 0.6991
Epoch 2/25: Validation Loss: 0.6910
Epoch 3/25: Validation Loss: 0.6834
Epoch 4/25: Validation Loss: 0.6918
Epoch 5/25: Validation Loss: 0.6943
Epoch 6/25: Validation Loss: 0.6896
Epoch 7/25: Validation Loss: 0.6864
Epoch 8/25: Validation Loss: 0.6900
Epoch 9/25: Validation Loss: 0.6954
Epoch 10/25: Validation Loss: 0.6928
Epoch 11/25: Validation Loss: 0.6916
Epoch 12/25: Validation Loss: 0.6970
Epoch 13/25: Validation Loss: 0.6973
Epoch 14/25: Validation Loss: 0.6957
Epoch 15/25: Validation Loss: 0.6968
Epoch 16/25: Validation Loss: 0.6944
Epoch 17/25: Validation Loss: 0.6918
Epoch 18/25: Validation Loss: 0.6982
Epoch 19/25: Validation Loss: 0.6920
Epoch 20/25: Validation Loss: 0.6932
Epoch 21/25: Validation Loss: 0.6947
Epoch 22/

[I 2024-10-03 20:20:16,603] Trial 19 finished with value: 0.6833597183227539 and parameters: {'learning_rate': 0.00044790758123736634, 'momentum': 0.9008586002953491, 'triplet_margin': 1.6185130873211857, 'classification_loss_weight': 0.8356618591942357}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.6935

Trial 20:
Suggested Hyperparameters: learning_rate=0.018410993659296073, momentum=0.8153277196709408, triplet_margin=1.2718528345638136, classification_loss_weight=0.9744735711039741
Epoch 1/25: Validation Loss: 0.7028
Epoch 2/25: Validation Loss: 0.6886
Epoch 3/25: Validation Loss: 0.6983
Epoch 4/25: Validation Loss: 0.8072
Epoch 5/25: Validation Loss: 0.6941
Epoch 6/25: Validation Loss: 0.6524
Epoch 7/25: Validation Loss: 0.6600
Epoch 8/25: Validation Loss: 0.6404
Epoch 9/25: Validation Loss: 0.7145
Epoch 10/25: Validation Loss: 0.7159
Epoch 11/25: Validation Loss: 0.8201
Epoch 12/25: Validation Loss: 0.6012
Epoch 13/25: Validation Loss: 0.5643
Epoch 14/25: Validation Loss: 0.5350
Epoch 15/25: Validation Loss: 0.5533
Epoch 16/25: Validation Loss: 0.6690
Epoch 17/25: Validation Loss: 0.4582
Epoch 18/25: Validation Loss: 0.6426
Epoch 19/25: Validation Loss: 0.5066
Epoch 20/25: Validation Loss: 0.3780
Epoch 21/25: Validation Loss: 0.5096
Epoch 22/25

[I 2024-10-03 20:22:16,356] Trial 20 finished with value: 0.3498817503452301 and parameters: {'learning_rate': 0.018410993659296073, 'momentum': 0.8153277196709408, 'triplet_margin': 1.2718528345638136, 'classification_loss_weight': 0.9744735711039741}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.5469

Trial 21:
Suggested Hyperparameters: learning_rate=0.0057372121931303964, momentum=0.6882593426478578, triplet_margin=1.1544877431752902, classification_loss_weight=0.9930864185680695
Epoch 1/25: Validation Loss: 0.6983
Epoch 2/25: Validation Loss: 0.6949
Epoch 3/25: Validation Loss: 0.7020
Epoch 4/25: Validation Loss: 0.7008
Epoch 5/25: Validation Loss: 0.7027
Epoch 6/25: Validation Loss: 0.6972
Epoch 7/25: Validation Loss: 0.6907
Epoch 8/25: Validation Loss: 0.6923
Epoch 9/25: Validation Loss: 0.6758
Epoch 10/25: Validation Loss: 0.6771
Epoch 11/25: Validation Loss: 0.6427
Epoch 12/25: Validation Loss: 0.6029
Epoch 13/25: Validation Loss: 0.6053
Epoch 14/25: Validation Loss: 0.5375
Epoch 15/25: Validation Loss: 0.5434
Epoch 16/25: Validation Loss: 0.5219
Epoch 17/25: Validation Loss: 0.4813
Epoch 18/25: Validation Loss: 0.4609
Epoch 19/25: Validation Loss: 0.4627
Epoch 20/25: Validation Loss: 0.4331
Epoch 21/25: Validation Loss: 0.4985
Epoch 22/2

[I 2024-10-03 20:24:13,684] Trial 21 finished with value: 0.3843324720859528 and parameters: {'learning_rate': 0.0057372121931303964, 'momentum': 0.6882593426478578, 'triplet_margin': 1.1544877431752902, 'classification_loss_weight': 0.9930864185680695}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.4264

Trial 22:
Suggested Hyperparameters: learning_rate=0.0018539274952970456, momentum=0.7349933431199192, triplet_margin=1.4239254118193532, classification_loss_weight=0.8864836313197098
Epoch 1/25: Validation Loss: 0.6967
Epoch 2/25: Validation Loss: 0.6983
Epoch 3/25: Validation Loss: 0.6984
Epoch 4/25: Validation Loss: 0.6960
Epoch 5/25: Validation Loss: 0.7015
Epoch 6/25: Validation Loss: 0.7001
Epoch 7/25: Validation Loss: 0.6991
Epoch 8/25: Validation Loss: 0.6964
Epoch 9/25: Validation Loss: 0.7008
Epoch 10/25: Validation Loss: 0.6924
Epoch 11/25: Validation Loss: 0.6956
Epoch 12/25: Validation Loss: 0.6967
Epoch 13/25: Validation Loss: 0.7013
Epoch 14/25: Validation Loss: 0.6937
Epoch 15/25: Validation Loss: 0.6942
Epoch 16/25: Validation Loss: 0.6923
Epoch 17/25: Validation Loss: 0.6796
Epoch 18/25: Validation Loss: 0.6806
Epoch 19/25: Validation Loss: 0.6757
Epoch 20/25: Validation Loss: 0.6759
Epoch 21/25: Validation Loss: 0.6693
Epoch 22/2

[I 2024-10-03 20:26:11,000] Trial 22 finished with value: 0.619137418270111 and parameters: {'learning_rate': 0.0018539274952970456, 'momentum': 0.7349933431199192, 'triplet_margin': 1.4239254118193532, 'classification_loss_weight': 0.8864836313197098}. Best is trial 10 with value: 0.2511402040719986.


Epoch 25/25: Validation Loss: 0.6191

Trial 23:
Suggested Hyperparameters: learning_rate=0.014669119460278886, momentum=0.6390119977051697, triplet_margin=1.129482285774702, classification_loss_weight=0.9139816288610945
Epoch 1/25: Validation Loss: 0.6996
Epoch 2/25: Validation Loss: 0.6902
Epoch 3/25: Validation Loss: 0.7056
Epoch 4/25: Validation Loss: 0.6854
Epoch 5/25: Validation Loss: 0.6764
Epoch 6/25: Validation Loss: 0.6572
Epoch 7/25: Validation Loss: 0.7056
Epoch 8/25: Validation Loss: 0.5469
Epoch 9/25: Validation Loss: 0.5870
Epoch 10/25: Validation Loss: 0.4524
Epoch 11/25: Validation Loss: 0.3974
Epoch 12/25: Validation Loss: 0.5240
Epoch 13/25: Validation Loss: 0.5190
Epoch 14/25: Validation Loss: 0.3866
Epoch 15/25: Validation Loss: 0.4858
Epoch 16/25: Validation Loss: 0.3551
Epoch 17/25: Validation Loss: 0.4237
Epoch 18/25: Validation Loss: 0.4653
Epoch 19/25: Validation Loss: 0.3417
Epoch 20/25: Validation Loss: 0.4864
Epoch 21/25: Validation Loss: 0.2539
Epoch 22/25:

[I 2024-10-03 20:28:08,522] Trial 23 finished with value: 0.2414734899997711 and parameters: {'learning_rate': 0.014669119460278886, 'momentum': 0.6390119977051697, 'triplet_margin': 1.129482285774702, 'classification_loss_weight': 0.9139816288610945}. Best is trial 23 with value: 0.2414734899997711.


Epoch 25/25: Validation Loss: 0.4841

Trial 24:
Suggested Hyperparameters: learning_rate=0.029773983149537487, momentum=0.6551874788829093, triplet_margin=0.8066913662269339, classification_loss_weight=0.8958748507133959
Epoch 1/25: Validation Loss: 0.7012
Epoch 2/25: Validation Loss: 0.7001
Epoch 3/25: Validation Loss: 0.6886
Epoch 4/25: Validation Loss: 0.6789
Epoch 5/25: Validation Loss: 0.7468
Epoch 6/25: Validation Loss: 0.6840
Epoch 7/25: Validation Loss: 0.6900
Epoch 8/25: Validation Loss: 0.5972
Epoch 9/25: Validation Loss: 0.6074
Epoch 10/25: Validation Loss: 0.5910
Epoch 11/25: Validation Loss: 0.5108
Epoch 12/25: Validation Loss: 0.5223
Epoch 13/25: Validation Loss: 0.4588
Epoch 14/25: Validation Loss: 0.5158
Epoch 15/25: Validation Loss: 0.5405
Epoch 16/25: Validation Loss: 0.5668
Epoch 17/25: Validation Loss: 0.3978
Epoch 18/25: Validation Loss: 0.3517
Epoch 19/25: Validation Loss: 0.5533
Epoch 20/25: Validation Loss: 0.5190
Epoch 21/25: Validation Loss: 0.4833
Epoch 22/25

[I 2024-10-03 20:30:05,246] Trial 24 finished with value: 0.3401291072368622 and parameters: {'learning_rate': 0.029773983149537487, 'momentum': 0.6551874788829093, 'triplet_margin': 0.8066913662269339, 'classification_loss_weight': 0.8958748507133959}. Best is trial 23 with value: 0.2414734899997711.


Epoch 25/25: Validation Loss: 0.6887

Trial 25:
Suggested Hyperparameters: learning_rate=0.012138312850152934, momentum=0.5003229954566906, triplet_margin=1.7439242795944683, classification_loss_weight=0.7152063810890267
Epoch 1/25: Validation Loss: 0.6851
Epoch 2/25: Validation Loss: 0.6914
Epoch 3/25: Validation Loss: 0.6926
Epoch 4/25: Validation Loss: 0.6915
Epoch 5/25: Validation Loss: 0.6983
Epoch 6/25: Validation Loss: 0.6905
Epoch 7/25: Validation Loss: 0.6879
Epoch 8/25: Validation Loss: 0.6780
Epoch 9/25: Validation Loss: 0.6581
Epoch 10/25: Validation Loss: 0.6284
Epoch 11/25: Validation Loss: 0.5622
Epoch 12/25: Validation Loss: 0.5426
Epoch 13/25: Validation Loss: 0.4931
Epoch 14/25: Validation Loss: 0.6328
Epoch 15/25: Validation Loss: 0.4542
Epoch 16/25: Validation Loss: 0.4607
Epoch 17/25: Validation Loss: 0.5010
Epoch 18/25: Validation Loss: 0.5525
Epoch 19/25: Validation Loss: 0.5662
Epoch 20/25: Validation Loss: 0.2783
Epoch 21/25: Validation Loss: 0.3756
Epoch 22/25

[I 2024-10-03 20:32:05,632] Trial 25 finished with value: 0.27825742959976196 and parameters: {'learning_rate': 0.012138312850152934, 'momentum': 0.5003229954566906, 'triplet_margin': 1.7439242795944683, 'classification_loss_weight': 0.7152063810890267}. Best is trial 23 with value: 0.2414734899997711.


Epoch 25/25: Validation Loss: 0.3292

Trial 26:
Suggested Hyperparameters: learning_rate=0.040272637957445345, momentum=0.5329360071948009, triplet_margin=1.6401995822091153, classification_loss_weight=0.709761057645813
Epoch 1/25: Validation Loss: 0.7525
Epoch 2/25: Validation Loss: 0.7278
Epoch 3/25: Validation Loss: 0.6923
Epoch 4/25: Validation Loss: 0.6971
Epoch 5/25: Validation Loss: 0.7335
Epoch 6/25: Validation Loss: 0.7224
Epoch 7/25: Validation Loss: 0.6649
Epoch 8/25: Validation Loss: 0.6753
Epoch 9/25: Validation Loss: 0.6789
Epoch 10/25: Validation Loss: 0.6083
Epoch 11/25: Validation Loss: 0.6282
Epoch 12/25: Validation Loss: 0.6344
Epoch 13/25: Validation Loss: 0.7128
Epoch 14/25: Validation Loss: 0.5301
Epoch 15/25: Validation Loss: 0.6172
Epoch 16/25: Validation Loss: 0.5208
Epoch 17/25: Validation Loss: 0.5468
Epoch 18/25: Validation Loss: 0.6439
Epoch 19/25: Validation Loss: 0.5747
Epoch 20/25: Validation Loss: 0.4401
Epoch 21/25: Validation Loss: 0.5894
Epoch 22/25:

[I 2024-10-03 20:34:07,481] Trial 26 finished with value: 0.4400989294052124 and parameters: {'learning_rate': 0.040272637957445345, 'momentum': 0.5329360071948009, 'triplet_margin': 1.6401995822091153, 'classification_loss_weight': 0.709761057645813}. Best is trial 23 with value: 0.2414734899997711.


Epoch 25/25: Validation Loss: 0.5271

Trial 27:
Suggested Hyperparameters: learning_rate=0.014803213977973965, momentum=0.5079861758547359, triplet_margin=1.703068819656492, classification_loss_weight=0.6232609307249704
Epoch 1/25: Validation Loss: 0.7000
Epoch 2/25: Validation Loss: 0.6993
Epoch 3/25: Validation Loss: 0.6867
Epoch 4/25: Validation Loss: 0.6875
Epoch 5/25: Validation Loss: 0.6951
Epoch 6/25: Validation Loss: 0.6936
Epoch 7/25: Validation Loss: 0.6708
Epoch 8/25: Validation Loss: 0.6185
Epoch 9/25: Validation Loss: 0.5958
Epoch 10/25: Validation Loss: 0.5661
Epoch 11/25: Validation Loss: 0.5454
Epoch 12/25: Validation Loss: 0.5496
Epoch 13/25: Validation Loss: 0.4846
Epoch 14/25: Validation Loss: 0.4507
Epoch 15/25: Validation Loss: 0.7294
Epoch 16/25: Validation Loss: 0.3214
Epoch 17/25: Validation Loss: 0.3278
Epoch 18/25: Validation Loss: 0.3741
Epoch 19/25: Validation Loss: 0.3036
Epoch 20/25: Validation Loss: 0.3827
Epoch 21/25: Validation Loss: 0.3317
Epoch 22/25:

[I 2024-10-03 20:36:08,516] Trial 27 finished with value: 0.17358753234148025 and parameters: {'learning_rate': 0.014803213977973965, 'momentum': 0.5079861758547359, 'triplet_margin': 1.703068819656492, 'classification_loss_weight': 0.6232609307249704}. Best is trial 27 with value: 0.17358753234148025.


Epoch 25/25: Validation Loss: 0.1736

Trial 28:
Suggested Hyperparameters: learning_rate=0.0029585645690902575, momentum=0.5769189560176743, triplet_margin=1.1490398651250644, classification_loss_weight=0.6464878726686589
Epoch 1/25: Validation Loss: 0.6937
Epoch 2/25: Validation Loss: 0.6892
Epoch 3/25: Validation Loss: 0.6950
Epoch 4/25: Validation Loss: 0.6953
Epoch 5/25: Validation Loss: 0.6902
Epoch 6/25: Validation Loss: 0.6867
Epoch 7/25: Validation Loss: 0.6901
Epoch 8/25: Validation Loss: 0.6969
Epoch 9/25: Validation Loss: 0.6909
Epoch 10/25: Validation Loss: 0.6947
Epoch 11/25: Validation Loss: 0.6932
Epoch 12/25: Validation Loss: 0.6916
Epoch 13/25: Validation Loss: 0.6917
Epoch 14/25: Validation Loss: 0.6970
Epoch 15/25: Validation Loss: 0.6882
Epoch 16/25: Validation Loss: 0.6865
Epoch 17/25: Validation Loss: 0.6853
Epoch 18/25: Validation Loss: 0.6868
Epoch 19/25: Validation Loss: 0.6812
Epoch 20/25: Validation Loss: 0.6729
Epoch 21/25: Validation Loss: 0.6639
Epoch 22/2

[I 2024-10-03 20:38:06,601] Trial 28 finished with value: 0.6495035886764526 and parameters: {'learning_rate': 0.0029585645690902575, 'momentum': 0.5769189560176743, 'triplet_margin': 1.1490398651250644, 'classification_loss_weight': 0.6464878726686589}. Best is trial 27 with value: 0.17358753234148025.


Epoch 25/25: Validation Loss: 0.6504

Trial 29:
Suggested Hyperparameters: learning_rate=0.02766420584125575, momentum=0.7414515378501716, triplet_margin=1.5156401875457117, classification_loss_weight=0.4709551861759709
Epoch 1/25: Validation Loss: 0.7007
Epoch 2/25: Validation Loss: 0.7069
Epoch 3/25: Validation Loss: 0.7107
Epoch 4/25: Validation Loss: 0.6919
Epoch 5/25: Validation Loss: 0.6928
Epoch 6/25: Validation Loss: 0.6966
Epoch 7/25: Validation Loss: 0.6898
Epoch 8/25: Validation Loss: 0.6904
Epoch 9/25: Validation Loss: 0.6887
Epoch 10/25: Validation Loss: 0.6924
Epoch 11/25: Validation Loss: 0.6880
Epoch 12/25: Validation Loss: 0.6781
Epoch 13/25: Validation Loss: 0.6720
Epoch 14/25: Validation Loss: 0.6767
Epoch 15/25: Validation Loss: 0.6610
Epoch 16/25: Validation Loss: 0.6370
Epoch 17/25: Validation Loss: 0.6445
Epoch 18/25: Validation Loss: 0.6583
Epoch 19/25: Validation Loss: 0.6758
Epoch 20/25: Validation Loss: 0.6601
Epoch 21/25: Validation Loss: 0.6464
Epoch 22/25:

[I 2024-10-03 20:40:06,275] Trial 29 finished with value: 0.6257834553718566 and parameters: {'learning_rate': 0.02766420584125575, 'momentum': 0.7414515378501716, 'triplet_margin': 1.5156401875457117, 'classification_loss_weight': 0.4709551861759709}. Best is trial 27 with value: 0.17358753234148025.


Epoch 25/25: Validation Loss: 0.6516
Best hyperparameters:  {'learning_rate': 0.014803213977973965, 'momentum': 0.5079861758547359, 'triplet_margin': 1.703068819656492, 'classification_loss_weight': 0.6232609307249704}
Best validation loss:  0.17358753234148025


Best hyperparameters:  {'learning_rate': 0.014803213977973965, 'momentum': 0.5079861758547359, 'triplet_margin': 1.703068819656492, 'classification_loss_weight': 0.6232609307249704}
Best validation loss:  0.17358753234148025